### ISSUE TITLE
Use stronger encryption method

### ISSUE THREAD

**Issue by riyadhalnur** [BosNaufal/secure-local-storage#1]
Looking over the code, the encryption can be easily compromised using brute force and it won't take much time or power. If you intend for people to use this in production systems, you need to use a stronger encryption. Just my 2 cents.

**Comment by BosNaufal** [comment]
Thank You for giving me a security hole issue. Would you like to prove me that you can get the data while you haven't the password? just for make sure. 
Thanks in advance~

**Comment by riyadhalnur** [comment]
I don't think I'll have the time to brute force your code but a recent news piece might interest you if you haven't seen it yet. A security researcher was able to break in to his facebook account using brute force due to a limited login attempt feature being turned off in Facebook BETA. He did not brute force it using a supercomputer and it's assumed in these cases that the attacker has no knowledge of the actual password.

**Comment by mike1808** [comment]
You have used the world's ancient encryption method which is called [Caesar cipher](https://en.m.wikipedia.org/wiki/Caesar_cipher)

It's very easy to brake.

**Comment by mike1808** [comment]
I recommend you to learn the basics of cryptography before creating libraries like this.

**Comment by BosNaufal** [comment]
Got your point! Thanks Again @riyadhalnur. It's a very good advice...
Thanks @mike1808 ... maybe I just too fast to call it "encryption". It's a good information that I have to read and learn. Thank you so much for your feedback

Every security system has a hole, but instead of let your data being plain and unsafe why not just using this. to make the others harder to do illegal inject into your data. 

Maybe will improve the encryption with other encryption library. Thanks all!

**Comment by riyadhalnur** [comment]
@BosNaufal Your idea is great but with cryptography, it's better to use battle hardened and tested encryption algorithms. Even some algorithms that were once thought to be strong enough are now obsolete because now they can be broken. 

My suggestion - modify the module to use AES-256 or the likes and not write your own encryption algorithm esp. if you want people to use this on live servers. :smiley:

**Comment by mike1808** [comment]
However, the main point is why you ever will need to encrypt your local storage on the client? It's [security through obscurity](https://en.wikipedia.org/wiki/Security_through_obscurity). Better to provide already encrypted data from the server (like tokens and other sensitive data).

**Comment by BosNaufal** [comment]
Whoooaaa, I'm feeling so excited about that! Will learn your suggestions then make it better~
I'll never know about that if there's nobody tell me. Thank you @riyadhalnur for apreciating my work. 

@mike1808, I have used it, but.. in some case (in SPA) there is a client want to minize the data traffic so I decided to store the role of user in the my localstorage instead of keep requesting to the server just for find out "the roles of the user". Please give me an advice if you have better solution~

**Comment by amovah** [comment]
Hi,

It may be good idea to use [crypto-js](https://github.com/brix/crypto-js) library. we can also add a option to permit user to choose a encryption method that want.

@BosNaufal what do think about ?

**Comment by BosNaufal** [comment]
Hi @amovah, that is what I've been looking for. Let's make an integration for it.

**Comment by amovah** [comment]
I'm familiar with crypto-js. If you want, we can do it together.

**Comment by BosNaufal** [comment]
Sure, Why not. 
I'm just a newbie here. So, I'll really need a help~ :grin:

**Comment by amovah** [comment]
So, are you agree to use crypto-js instead your own encryption method?

If you agree, I'm interested in to work on it.

**Comment by mike1808** [comment]
In crypto.js they have a very handy abstract class which is called `BlockCipher`. You can support these class methods and any encryption which implements those methods can be used with your library. 
And remember the main rule using cryptography: "Never ever implement your own encryption algorithm".

**Comment by BosNaufal** [comment]
Yeah Sure @amovah, I'm not an idealist person like that. So, instead being doubt with it, why not just change it to another tested library. Let's do it :grinning: 

Thanks for the information @mike1808. Will find out it more and I will keep in mind to "Never ever implement your own encryption algorithm". I learn so much from you, thank you~

**Comment by amovah** [comment]
@BosNaufal , what is your suggestion to use crypto-js in our library? using [webpack](https://webpack.github.io/) or demand users to load crypto as a dependency?

**Comment by BosNaufal** [comment]
Hmm.... How about using booth of them? @amovah

**Comment by amovah** [comment]
So, we should provide two version, one of them with crypto inside itself and another one without crypto.

hmm.. If you agree, we can start.

**Comment by BosNaufal** [comment]
I do Agree with you~ Let's start to develope it :grin: 
I'm not familiar with Crypto-JS so I have no idea to use it. That's why I'm waiting for your branch update then I will find out, what I can do. 
For the moment, I've just worked to add some new features of it.

**Comment by amovah** [comment]
@BosNaufal , OK. I will make environment for that ASAP.

**Comment by osamasaeed** [comment]
@BosNaufal  Your great work I really like it.
please cloud you make a module for angular 2x
